include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,230 @@
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        FAIL_FAST_IF(storage == details::s_apartmentStorage.end());
+        auto& variables = storage->second;
+        variables.erase(variableKey);
+    }
+
+    // If the variable exists, replace it with a new value. This has no
+    // effect the variable was not stored, is this too loose of a contract?
+    template<typename F>
+    void reset_for_current_com_apartment(const F& createFn, std::invoke_result_t<F> newValue)
+    {
+        static_assert(std::is_function_v<F>, "createFn must be a function");
+
+        unsigned long long apartmentId{};
+        FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+
+        const auto variableKey = std::addressof(createFn);
+
+        std::any newAny(newValue); // release outside of the lock
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        if (storage != details::s_apartmentStorage.end())
+        {
+            auto& variables = storage->second;
+            auto variable = variables.find(variableKey);
+            if (variable != variables.end())
+            {
+                variable->second.swap(newAny);
+            }
+        }
+    }
+
+    // Replace a variable with a new value. This requires (ensured via fail fast) that the
+    // variable is already stored based on a call to get_for_current_com_apartment in the
+    // current apartment.
+    template<typename F>
+    void reset_for_current_com_apartment_strict(const F& createFn, std::invoke_result_t<F> newValue)
+    {
+        static_assert(std::is_function_v<F>, "createFn must be a function");
+
+        unsigned long long apartmentId{};
+        FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+
+        const auto variableKey = std::addressof(createFn);
+
+        std::any newAny(newValue); // release outside of the lock
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        FAIL_FAST_IF(storage == details::s_apartmentStorage.end());
+        auto& variables = storage->second;
+        auto variable = variables.find(variableKey);
+        FAIL_FAST_IF(variable == variables.end());
+        variable->second.swap(newAny);
+    }
+
+    // for testing
+    size_t apartment_count() { return details::s_apartmentStorage.size(); }
+    size_t apartment_variable_count()
+    {
+        unsigned long long apartmentId{};
+        FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        if (storage != details::s_apartmentStorage.end())
+        {
+            return storage->second.size();
+        }
+        return 0;
+    }
+}
+
+#endif // __WIL_COM_APARTMENT_VARIABLE_INCLUDED

```

Comment:
Here's another pattern: The `get` function returns a reference to the storage, not a copy. (We might even default to using the constructor instead of a custom maker function.)

```cpp
template<typename Storage>
auto& get_for_current_com_apartment()
{
   add an entry if necessary
   return a reference to the s_apartmentStorage[apartmentId][key];
}
```

The `try` version returns a pointer to the storage, or `nullptr` if it doesn't exist.
```cpp
template<typename Storage>
auto try_get_for_current_com_apartment()
{
    if (no entry exists) return nullptr;
    else return a pointer to the s_apartmentStorage[apartmentId][key];
}

```

(Returning a reference or pointer to the value storage is legal because `unordered_map` guarantees that the only thing that invalidates a key or value reference/pointer is the erasure of that key/value pair.)

This allows you to use non-copyable types with `get_for_current_com_apartment`. It also lets you do things like

```cpp
    // another widget for this thread
    ++wil::get_for_current_com_apartment<WidgetCounterStorage>();
```

or

```cpp
    // if we are the current widget, then say that there is no current widget any more
    auto currentWidget = wil::try_get_for_current_com_apartment<std::shared_ptr<Widget>>();
    if (currentWidget && *currentWidget == this) *currentWidget = nullptr;
```

Reply:
I took your first suggestion. on the second, my thoughts are callers that want to support empty will store a null value and treat it specially. My intended use for this feature will be to store the crawl scope manager and I plan to store the HRESULT from the attempts to get it and the pointer, if retrieved. that will enable treating the cases where the crawl scope manager can't be retrieved due to the service being disabled distinctly from other failures (out of memory).